File: apps/idenplane/packages/idenplane-android/src/main/kotlin/com/idenplane/sdk/PKCEHelper.kt
package com.idenplane.sdk

import android.util.Base64
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * Helpers for generating OAuth 2.0 PKCE parameters (RFC 7636).
 *
 * Generates a cryptographically random code verifier and derives
 * the S256 code challenge via SHA-256.
 */
object PKCEHelper {

    private val secureRandom = SecureRandom()

    // -----------------------------------------------------------------------
    // Code Verifier
    // -----------------------------------------------------------------------

    /**
     * Generate a cryptographically random PKCE code verifier.
     *
     * The verifier is 32 random bytes encoded as Base64URL without padding,
     * producing a 43-character string within the RFC 7636 §4.1 bounds (43–128).
     */
    fun generateCodeVerifier(): String {
        val bytes = ByteArray(32)
        secureRandom.nextBytes(bytes)
        return bytes.base64UrlEncode()
    }

    // -----------------------------------------------------------------------
    // Code Challenge
    // -----------------------------------------------------------------------

    /**
     * Derive the S256 PKCE code challenge from a verifier string.
     *
     * Challenge = BASE64URL(SHA256(ASCII(verifier)))
     */
    fun generateCodeChallenge(verifier: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val hash   = digest.digest(verifier.toByteArray(Charsets.US_ASCII))
        return hash.base64UrlEncode()
    }

    // -----------------------------------------------------------------------
    // State
    // -----------------------------------------------------------------------

    /**
     * Generate a random state parameter for CSRF protection.
     */
    fun generateState(): String {
        val bytes = ByteArray(16)
        secureRandom.nextBytes(bytes)
        return bytes.base64UrlEncode()
    }

    // -----------------------------------------------------------------------
    // Internal helpers
    // -----------------------------------------------------------------------

    private fun ByteArray.base64UrlEncode(): String =
        Base64.encodeToString(this, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
}

File: apps/idenplane/packages/idenplane-ios/Sources/Idenplane/PKCEHelper.swift
import Foundation
import CryptoKit

/// Helpers for generating OAuth 2.0 PKCE parameters (RFC 7636).
public enum PKCEHelper {

    // MARK: - Code Verifier

    /// Generate a cryptographically random PKCE code verifier.
    ///
    /// The verifier is a high-entropy random string of 43–128 characters,
    /// Base64URL-encoded without padding, as required by RFC 7636.
    public static func generateCodeVerifier() -> String {
        var buffer = [UInt8](repeating: 0, count: 32)
        let result = SecRandomCopyBytes(kSecRandomDefault, buffer.count, &buffer)
        guard result == errSecSuccess else {
            // Fallback to Swift's built-in random if SecRandom fails
            buffer = (0..<32).map { _ in UInt8.random(in: 0...255) }
            return Data(buffer).base64URLEncodedString()
        }
        return Data(buffer).base64URLEncodedString()
    }

    // MARK: - Code Challenge

    /// Derive the S256 PKCE code challenge from a verifier.
    ///
    /// Challenge = BASE64URL(SHA256(ASCII(verifier)))
    public static func generateCodeChallenge(from verifier: String) -> String {
        guard let data = verifier.data(using: .ascii) else {
            preconditionFailure("Code verifier must be ASCII")
        }
        let digest = SHA256.hash(data: data)
        return Data(digest).base64URLEncodedString()
    }

    // MARK: - State

    /// Generate a random state parameter for CSRF protection.
    public static func generateState() -> String {
        var buffer = [UInt8](repeating: 0, count: 16)
        _ = SecRandomCopyBytes(kSecRandomDefault, buffer.count, &buffer)
        return Data(buffer).base64URLEncodedString()
    }
}

// MARK: - Data + Base64URL

private extension Data {
    /// Encode as Base64URL without padding characters (RFC 4648 §5).
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
